paasta_tools/cli/cmds/spark_run.py

```diff
@@ -513,25 +513,33 @@ def get_docker_run_cmd(
 def get_docker_image(
     args: argparse.Namespace, instance_config: InstanceConfig
 ) -> Optional[str]:
+    """
+    Since the Docker image digest used to launch the Spark cluster is obtained by inspecting local
+    Docker images, we need to ensure that the Docker image exists locally or is pulled in all scenarios.
+    """
+    # docker image is built locally then pushed
     if args.build:
         return build_and_push_docker_image(args)
+
+    docker_url = ""
     if args.image:
-        return args.image
+        docker_url = args.image
+    else:
+        try:
+            docker_url = instance_config.get_docker_url()
+        except NoDockerImageError:
+            print(
+                PaastaColors.red(
+                    "Error: No sha has been marked for deployment for the %s deploy group.\n"
+                    "Please ensure this service has either run through a jenkins pipeline "
+                    "or paasta mark-for-deployment has been run for %s\n"
+                    % (instance_config.get_deploy_group(), args.service)
+                ),
+                sep="",
+                file=sys.stderr,
+            )
+            return None
 
-    try:
-        docker_url = instance_config.get_docker_url()
-    except NoDockerImageError:
-        print(
-            PaastaColors.red(
-                "Error: No sha has been marked for deployment for the %s deploy group.\n"
-                "Please ensure this service has either run through a jenkins pipeline "
-                "or paasta mark-for-deployment has been run for %s\n"
-                % (instance_config.get_deploy_group(), args.service)
-            ),
-            sep="",
-            file=sys.stderr,
-        )
-        return None
```

Comment:
In Line548, https://github.com/Yelp/paasta/pull/3614/files#diff-30f8e9c33224158fab1e77efc4004f7d2344230ea7b4ffbf64b3fca21513cc00R548 should we be checking if docker image exists locally befor pulling? Or the check is happening implicitly?

Reply:
For the case of the local image is newer than the one on registry (built, tagged locally but didn't push), `docker pull` will always pull the latest version on the registry anyway.

More details:
https://github.com/moby/moby/issues/10291#issuecomment-71655034